### PR TITLE
[assets] add partition_def and config to define_asset_job

### PR DESCRIPTION
**Pull request by OwenKephart**
### Summary & Motivation

want to support these use cases

for the partition bit, depends on: https://github.com/dagster-io/dagster/pull/8245

### How I Tested These Changes

pretty flimsy unit tests

### CODE REVIEW

File: python_modules/dagster/dagster/core/definitions/unresolved_asset_job_definition.py [hunk 1]
```diff
@@ -78,22 +107,43 @@ def _selection_from_string(string: str) -> "AssetSelection":
...
     tags: Optional[Dict[str, Any]] = None,
+    partitions_def: Optional["PartitionsDefinition"] = None,
 ) -> UnresolvedAssetJobDefinition:
     """Creates a definition of a job which will materialize a selection of assets. This will only be
     resolved to a JobDefinition once placed in a repository.
 
     Args:
         name (str):
-            The name for the JobDefintion.
+            The name for the Job.
```
Comment (sryza) [inline python_modules/dagster/dagster/core/definitions/unresolved_asset_job_definition.py:120]: A little weird to capitalize Job?

File: python_modules/dagster/dagster/core/definitions/unresolved_asset_job_definition.py [hunk 2]
```diff
@@ -78,22 +107,43 @@ def _selection_from_string(string: str) -> "AssetSelection":
...
+            If a dictionary is provided, then it must conform to the standard config schema, and
+            it will be used as the job's run config for the job whenever the job is executed.
+            The values provided will be viewable and editable in the Dagit playground, so be
+            careful with secrets.
+
+            If a :py:class:`ConfigMapping` object is provided, then the schema for the job's run config is
+            determined by the config mapping, and the ConfigMapping, which should return
+            configuration in the standard format to configure the job.
+
+            If a :py:class:`PartitionedConfig` object is provided, then it defines a discrete set of config
```
Comment (sryza) [inline python_modules/dagster/dagster/core/definitions/unresolved_asset_job_definition.py:143]: It hurts my brain a little bit to think about how PartitionedConfig would work with partitioned assets. Maybe we should leave this out for now?

File: python_modules/dagster/dagster/core/definitions/unresolved_asset_job_definition.py [hunk 3]
```diff
@@ -78,22 +107,43 @@ def _selection_from_string(string: str) -> "AssetSelection":
...
+            values that can parameterize the job, as well as a function for mapping those
+            values to the base config. The values provided will be viewable and editable in the
+            Dagit playground, so be careful with secrets.
```
Comment (sryza) [inline python_modules/dagster/dagster/core/definitions/unresolved_asset_job_definition.py:146]: Worth adding docstring for `partitions_def`. And, in it, documenting any constraints (all assets selected in the job need to have that partitioning?)

**Review by sryza — COMMENTED** [review]
Nice

File: python_modules/dagster/dagster/core/definitions/unresolved_asset_job_definition.py (hunk 2 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster/core/definitions/unresolved_asset_job_definition.py:143]: yeah my brain broke trying to consider that as well. I'll leave it out of the docstring

**Review by sryza — APPROVED** [review]

**Pull request merged**